Add modified DSIR scripts for general use cases

tianxie-30 opened the pull request:
I was using a machine with many CPU cores and large size of memory to run DSIR.

I split the data pool into many files with similar sizes, also split the target data. The parallelism is on file-level.

I didn't chunk the data here. If the data pool has many different mixed sources, like the Pile, we need to first split the data into different subsets (CommonCrawl, Wikipedia, books, etc) or chunk the data.

Commit:
add modified DSIR scripts

Commit:
time stamp is optional

sangmichaelxie committed:
DSIR and HashedNgramDSIR classes

sangmichaelxie committed:
create dir in super

sangmichaelxie committed:
generalize path loading

sangmichaelxie commented on data_selection/dsir_general/data_selection.py:
```diff
@@ -0,0 +1,202 @@
+
+    nonzero_idxs = np.where(global_mask)[0]
+    # choose top k
+    chosen_idxs = np.argpartition(-logratios, args.num_to_retrieve)[:args.num_to_retrieve]
```
we can use chosen_idxs directly here (without indexing into nonzero_idxs), since global_mask isn't being modified by another filtering step

sangmichaelxie committed:
change package name to data_selection

sangmichaelxie committed:
virtual sharding

sangmichaelxie committed:
save log importance weights to disk

sangmichaelxie committed:
avoid loading json into dict

sangmichaelxie committed:
readme

sangmichaelxie commented:
> I was using a machine with many CPU cores and large size of memory to run DSIR.
> 
> I split the data pool into many files with similar sizes, also split the target data. The parallelism is on file-level.
> 
> I didn't chunk the data here. If the data pool has many different mixed sources, like the Pile, we need to first split the data into different subsets (CommonCrawl, Wikipedia, books, etc) or chunk the data.

I tested it without chunking the data or splitting the data into different subsets and it turned out to be fine. The reason seems to be that, if we have long documents that are not relevant, the irrelevant tokens will actually contribute negative log importance weight. (instead of accumulating a lot of small, positive weights). For relevant documents, longer documents with more relevant tokens will be preferred over short but relevant documents. The bad case is actually that some short irrelevant documents can end up in the selected documents since there aren't enough tokens that contribute negative weight and some tokens can get lucky and have positive weight due to hash collisions.

tianxie-30 commented:
> > I was using a machine with many CPU cores and large size of memory to run DSIR.
> > I split the data pool into many files with similar sizes, also split the target data. The parallelism is on file-level.
> > I didn't chunk the data here. If the data pool has many different mixed sources, like the Pile, we need to first split the data into different subsets (CommonCrawl, Wikipedia, books, etc) or chunk the data.
> 
> I tested it without chunking the data or splitting the data into different subsets and it turned out to be fine. The reason seems to be that, if we have long documents that are not relevant, the irrelevant tokens will actually contribute negative log importance weight. (instead of accumulating a lot of small, positive weights). For relevant documents, longer documents with more relevant tokens will be preferred over short but relevant documents. The bad case is actually that some short irrelevant documents can end up in the selected documents since there aren't enough tokens that contribute negative weight and some tokens can get lucky and have positive weight due to hash collisions.

Thank you @sangmichaelxie ! I also noticed similar results. The worst retrieved documents are the short irrelevant documents.

sangmichaelxie committed:
filtering by metadata

sangmichaelxie committed:
readme

sangmichaelxie reviewed (approved):
thanks @tianxie-sf for the headstart on the general implementation!

sangmichaelxie commented:
> > > I was using a machine with many CPU cores and large size of memory to run DSIR.
> > > I split the data pool into many files with similar sizes, also split the target data. The parallelism is on file-level.
> > > I didn't chunk the data here. If the data pool has many different mixed sources, like the Pile, we need to first split the data into different subsets (CommonCrawl, Wikipedia, books, etc) or chunk the data.
> > 
> > 
> > I tested it without chunking the data or splitting the data into different subsets and it turned out to be fine. The reason seems to be that, if we have long documents that are not relevant, the irrelevant tokens will actually contribute negative log importance weight. (instead of accumulating a lot of small, positive weights). For relevant documents, longer documents with more relevant tokens will be preferred over short but relevant documents. The bad case is actually that some short irrelevant documents can end up in the selected documents since there aren't enough tokens that contribute negative weight and some tokens can get lucky and have positive weight due to hash collisions.
> 
> Thank you @sangmichaelxie ! I also noticed similar results. The worst retrieved documents are the short irrelevant documents.

I ended up incorporating a default filter on length, excluding all documents with less than 100 tokens (you can also use whatever number you may want to specify). Hopefully that is enough to get good statistics on most documents.
